internal/cobraprompt/prompt.go

```diff
@@ -151,54 +153,67 @@ func RegisterPersistFlag(co *cobra.Command) {
 }
 
 func findSuggestions(co *CobraPrompt, d *prompt.Document) []prompt.Suggest {
-	command := co.RootCmd
+	upToCursor := d.CurrentLineBeforeCursor()
+	// use line before cursor for command suggestion
+	bArgs := strings.Fields(upToCursor)
+	command, _, err := co.RootCmd.Find(bArgs)
+	if err != nil && strings.Contains(upToCursor, " ") {
+		return nil
+	}
+	wordBeforeCursor := d.GetWordBeforeCursor()
+	// use whole line for flag suggestions
 	args := strings.Fields(d.CurrentLine())
-	if found, _, err := command.Find(args); err == nil {
-		command = found
+	suggestions := traverseForFlagSuggestions(wordBeforeCursor, args, co, command)
+	if command.HasAvailableSubCommands() {
+		for _, c := range command.Commands() {
+			if !c.Hidden && !co.ShowHiddenCommands {
+				suggestions = append(suggestions, prompt.Suggest{Text: c.Name(), Description: c.Short})
+			}
+			if co.ShowHelpCommandAndFlags {
+				c.InitDefaultHelpFlag()
+			}
+		}
 	}
+	annotation := command.Annotations[DynamicSuggestionsAnnotation]
+	if co.DynamicSuggestionsFunc != nil && annotation != "" {
+		suggestions = append(suggestions, co.DynamicSuggestionsFunc(annotation, d)...)
+	}
+	return prompt.FilterHasPrefix(suggestions, wordBeforeCursor, true)
+}
+
+func traverseForFlagSuggestions(wordBeforeCursor string, words []string, co *CobraPrompt, command *cobra.Command) []prompt.Suggest {
 	var suggestions []prompt.Suggest
-	persistFlagValues, err := command.Flags().GetBool(PersistFlagValuesFlag)
-	if err != nil {
-		fmt.Println("cannot parse persist flag err: ", err)
+	noWordTyped := wordBeforeCursor == ""
+	dashPrefix := strings.HasPrefix(wordBeforeCursor, "-")
+	if !noWordTyped && !dashPrefix {
+		// no flag prefix
+		return suggestions
 	}
 	addFlags := func(flag *pflag.Flag) {
-		if flag.Changed && !persistFlagValues {
-			flag.Value.Set(flag.DefValue)
-		}
 		if flag.Hidden && !co.ShowHiddenFlags {
 			return
 		}
 		if stringInSlice(co.FlagsToExclude, flag.Name) {
 			return
 		}
 		flagUsage := "--" + flag.Name
-		if strings.HasPrefix(d.GetWordBeforeCursor(), "--") {
+		// Check if flag is already used in the command
+		if (flag.Shorthand != "" && internal.IsOneOf(words, "-"+flag.Shorthand, true)) ||
+			internal.IsOneOf(words, flagUsage, true) {
```

Comment:
How about introducing the `internal/check` package for `IsOneOf` just like we do for the Go client?

Reply:
Sounds good. Fixed at https://github.com/hazelcast/hazelcast-commandline-client/pull/41/commits/0bdfd09792e7af0022e3b8ab244bec7b2813bd57